R/teal_data-get_code.R

```diff
@@ -6,9 +6,62 @@
 #' Use `datanames` to limit the code to one or more of the data sets enumerated in `@datanames`.
 #' If the code has not passed verification (with [`verify()`]), a warning will be prepended.
 #'
-#' @section Notes for Developers:
-#' To learn more about how a subset of code needed to reproduce a specific data set is extracted from all code,
-#' see [`get_code_dependency()`].
+#' @section Extracting dataset-specific code:
+#' When `datanames` is specified `get_code` will limits the output only to the lines of code needed
+#' to recreate the requested data sets. `code` stored in the `teal_data` object is analyzed statically
+#' to determine dependency tree between each line of the code. Analysis is performed automatically
+#' based on the used symbols and it is working in a standard case when a new dataset is created by
+#' the assignment operator.
+#'
+#' Consider the following example:
+#' ```r
+#' data <- teal_data() |>
+#'   within(
+#'     foo <- function(x) x + 1
+#'     x <- 0
+#'     y <- foo(x)
+#'   )
+#' get_code(data, datanames = "y")
+#' ```
+#'
+#' In above case `y` depends on `x` and `foo` so the code returned by `get_code` will contain all
+#' three lines of code. `get_code(data, datanames = "x")` will return only the second line of code etc.
+#' \cr
+#' When a code uses non-standard evaluation `get_code` won't be able to determine relationships
```

Comment:
WDYT @chlebowa @m7pr

Reply:
I don't quite understand. Can you provide an example of NSE code that is not analyzed properly?

Reply:
Good point.
```
{ 
  env <- parent.frame()
  env$x <- 0
}
```
 I took NSE as a term explaining above but it doesn't seem to precise for me. NSE definition is also unwritten. It refers to code substitution and custom scoping rules. I'll try to find better words for this

Reply:
We meant code like that
```r
z <- 'a'
assign(z, 5)
foo <- function(x, value) {
    assign(x, value, envir = .GlobalEnv)
 }
 foo('a', 5)
```

Reply:
`library(magrittr)` is NSE, as opposed to `library("magrittr"). So is `list$element` as opposed to `lit[["element"]]`. `within` uses NSE as well but `x <- within(iris, Sepal.Length * Sepal.Width)` is processed properly.

Reply:
> We meant code like that

This is standard evaluation.

Reply:
Non-standard environment scoping, custom environment scoping - any other propositions to replace NSE?

Reply:
I'll change this line to: 
> `get_code` isn't always able to properly assess dependencies between each calls and symbols in a provided code.

I will not specify then what exactly causes parser to fail, because the list might be too long and still not exhaustive. What it doesn't support or might fail on:

- updating and reusing objects in a different environment (non standard scoping)
- NSE
- custom assign operators (like `%<>%`) etc.

 I wouldn't specify very technical terms either as 99% of the users will have problems to comprehend it

Reply:
> `get_code` isn't always able to properly assess dependencies between each calls and symbols in a provided code.

I think this is too broad. Here is a function that detects dependencies between symbols but it doesn't always work.

Reply:
The reason I asked the initial question is I cannot really say where the mechanism is expected to fail myself. The `env$x` example you provide is so extreme I cannot imagine a use case in the the context of preprocessing data for `teal` applications.

Reply:
Well, maybe we can say it analyzes the code and detects standard relations that result from `<-` operator? Object relations specified in different way may not work, hence we provided `#@linksto` tag to manually specify relations created with other way than `<-` operator.

Reply:
> analyzes the code and detects standard relations that result from `<-` operator? Object relations specified in different way may not work, hence we provided #@linksto tag to manually specify relations created with other way than `<-` operator.

Fine for me as it highlight what is properly parsed for sure.

Reply:
@chlebowa any counter proposition?

Reply:
Well, we do handle `assign` and `data`, and (I think) `<<-` too.

Sorry if this feels like nitpicking. I think I'm missing something: I don't know what the function is unable to do.

Reply:
@chlebowa 

Not able to determine dependencies on functions which have side effects

```r
data <- teal_data() |> eval_code(
  code = "
    x <- 0
    foo <- function() {
      x <<- x + 1
    }
    foo()
    y <- x
  "
)

identical(
  get_code(data, datanames = "x"),
  paste(
    "x <- 0",
    "foo <- function() {",
    "    x <<- x + 1",
    "}",
    "foo()",
    collapse = "\n"
  )
)
# [1] FALSE
```

not able to track "language" objects
```r
data <- teal_data() |> eval_code(
  code = "
    expr <- quote(x + 1)
    x <- 0
    eval(expr)
  "
)

identical(
  get_code(data, datanames = "x")
  paste(
    "expr <- quote(x + 1)",
    "x <- 0",
    "eval(expr)"
  )
)
# [1] FALSE
```

Doesn't understand options object
```
data <- teal_data() |> eval_code(
  code = '
    options(super_option = TRUE)
    if (getOption("super_option")) {
      x <- 0
    }
  '
)

identical(
  get_code(data, datanames = "x"),
  paste(
    "options(super_option = TRUE)",
    "if (options(\"super_option\")) {",
    "    x <- 0",
    "}",
    collapse = "\n"
  )
)
# [1] FALSE
```

understands when environment is an object (assumess it is a lists)
```r
data <- teal_data() |> eval_code(
  code = '
    y <- 0
    .GlobalEnv$hehe <- y
    x <- .GlobalEnv$hehe
  '
)

identical(
  get_code(data, datanames = "x"),
  paste(
    "y <- 0",
    ".GlobalEnv$hehe <- y",
    "x <- .GlobalEnv$hehe"
  )
)
```

Can't properly determine scoping of the functions
```r
data <- teal_data() |> eval_code(
  code = '
    Sepal.Length <- 0
    x <- dplyr::mutate(iris, yy = Sepal.Length)
  '
)

identical(
  get_code(data, datanames = "x"),
  paste(
    "x <- dplyr::mutate(iris, yy = Sepal.Length)"
  )
  )
)
# [1] FALSE
```

Can't conclude what ifluences environment and that does not
```r
data <- teal_data() |> eval_code(
  code = '
    x <- 1
    assign("x", 0, envir = environment())
    y <- x
  '
)

identical(
  get_code(data, datanames = "y"),
  paste(
    'x <- 1',
    'assign("x", 0, envir = environment())',
    'y <- x'
  )
)
# [1] FALSE
```

Reply:
Can't relate to for loops
```r
data <- teal_data() |> eval_code(
  code = '
    x <- 10
    y <- list()
    for (xx in 1:x) {
      x[[xx]] <- xx
    }
  '
)

identical(
  get_code(data, datanames = "y"),
  paste(
    "x <- 10",
    "y <- list()",
    "for (xx in 1:x) {",
    "    x[[xx]] <- xx",
    "}",
    collapse = "\n"
  )
)
# [1] FALSE
```

Reply:
All above can be linked by adding `# @linksto` of course

Reply:
Thank you, I can work with that :+1:

Reply:
> understands when environment is an object (assumess it is a lists)

This one actually works, you missed a `sep = "\n"`.

> Can't properly determine scoping of the functions

This one is tricky because the first line is included, where it should be omitted. We would need an ignore tag to handle this, I guess.
This is the only example of NSE, by the way.

> Can't conclude what ifluences environment and that does not

This is actually a bug because `data[["y"]]` ends up being `0` in both cases. @m7pr

Reply:
> > not able to track "language" objects

> There are some bugs here, which we have attempted to tackle but failed.

We didn't scope that initially. I created an issue for that, for the next release https://github.com/insightsengineering/teal.data/issues/270

> > Can't conclude what ifluences environment and that does not

> This is actually a bug because data[["y"]] ends up being 0 in both cases. @m7pr

Yeah, something is off as `assign("x", 0, envir = environment())` should be returned. Will start one more issue to handle that https://github.com/insightsengineering/teal.data/issues/271